### PR TITLE
Feature/create coupon crud and apply coupon

### PR DESCRIPTION
**Pull request by hfatima08**

### CODE REVIEW

File: app/Http/Controllers/CouponController.php [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+    // Injected Service 
+    public function __construct(PaymentServiceInterface $paymentService)
+    {
+        $this->paymentService = $paymentService;
+    }
+// -- to fetch all coupons with pagination --
+    public function index() {
+        try {
+           $coupons = Coupon::paginate($this->defaultPageSize);
```
Comment (ZeeshanHabib7) [inline app/Http/Controllers/CouponController.php]: make it as user fetch api

File: app/Http/Controllers/CouponController.php [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+     }
+
+ 
+     public function store(array $payload) {
+        try {
+            // Check if coupon is for product or cart
+            if (array_key_exists("product_id" ,$payload) && !is_null($payload['product_id'])) {
+              // ----- Coupon is for product -----
+                //get product price
+                $product = Product::find($payload["product_id"]);
```
Comment (ZeeshanHabib7) [inline app/Http/Controllers/CouponController.php]: what if product not found the code will crash on the very next line as you are using price from result set of product

Comment (ZeeshanHabib7) [reply app/Http/Controllers/CouponController.php]: use findOrfail here

File: app/Http/Controllers/CouponController.php [hunk 3]
```diff
@@ -97,7 +84,36 @@ public function update(array $payload, $id) {
      
 // -- Delete coupon --
      public function destroy($id) {
+        try {
+            $coupon = Coupon::find($id);
```
Comment (ZeeshanHabib7) [inline app/Http/Controllers/CouponController.php]: use findOrfail() here please

File: app/Http/Controllers/CouponController.php [hunk 4]
```diff
@@ -97,7 +84,36 @@ public function update(array $payload, $id) {
...
+            return handleException($e);
+        }
+     }
 
+// -- check if coupon is for product --
+     public function isProductCoupon($payload) {
+        try {
+            if (array_key_exists("product_id" ,$payload) && !is_null($payload['product_id'])) {
+                  //get product price
+                  $product = Product::find($payload["product_id"]);
```
Comment (ZeeshanHabib7) [inline app/Http/Controllers/CouponController.php]: here too

**Review by SyedSamrozeAli — COMMENTED** [review]
In first condition you have used "usage_limit_min" two times. Will this not give any error?

File: app/Http/Requests/CouponRequest.php
Comment (SyedSamrozeAli) [inline app/Http/Requests/CouponRequest.php:1]: You have written ""usage_limit_min" two times in the first condition. Will this not give any error?

**Pull request merged**